README.md
# Gatsby plugin Algolia

> This plugin is in _beta_ and not officially supported yet
>
> Feel free to open issues for any questions or ideas

You can specify a list of queries to run and how to transform them into an array of objects to index. When you run `gatsby build`, it will publish those to Algolia.

<div align="center">
  <a
    href="https://mermaid-js.github.io/mermaid-live-editor/edit##eyJjb2RlIjoiZ3JhcGggTFJcbiAgICBBW1NvdXJjZSAxXSAtLT4gfHF1ZXJ5fCBHYXRzYnlcbiAgICBCW1NvdXJjZSAyXSAtLT4gfHF1ZXJ5fCBHYXRzYnlcbiAgICBDW1NvdXJjZSAzXSAtLT4gfHF1ZXJ5fCBHYXRzYnlcbiAgICBcbiAgICBHYXRzYnkgLS0-IHxnYXRzYnkgYnVpbGR8IEFsZ29saWEiLCJtZXJtYWlkIjoie1xuICBcInRoZW1lXCI6IFwibmV1dHJhXCJcbn0iLCJ1cGRhdGVFZGl0b3IiOmZhbHNlLCJhdXRvU3luYyI6dHJ1ZSwidXBkYXRlRGlhZ3JhbSI6ZmFsc2V9"
  >
    <img
      src="https://mermaid.ink/svg/eyJjb2RlIjoiZ3JhcGggTFJcbiAgICBBW1NvdXJjZSAxXSAtLT4gfHF1ZXJ5fCBHYXRzYnlcbiAgICBCW1NvdXJjZSAyXSAtLT4gfHF1ZXJ5fCBHYXRzYnlcbiAgICBDW1NvdXJjZSAzXSAtLT4gfHF1ZXJ5fCBHYXRzYnlcbiAgICBcbiAgICBHYXRzYnkgLS0-IHxnYXRzYnkgYnVpbGR8IEFsZ29saWEiLCJtZXJtYWlkIjp7InRoZW1lIjoibmV1dHJhbCJ9LCJ1cGRhdGVFZGl0b3IiOmZhbHNlLCJhdXRvU3luYyI6dHJ1ZSwidXBkYXRlRGlhZ3JhbSI6ZmFsc2V9"
      alt="flow diagram"
    />
  </a>
</div>

Here we have an example with some data that might not be very relevant, but will work with the default configuration of `gatsby new`

```shell
yarn add gatsby-plugin-algolia
```

First add credentials to a .env file, which you won't commit. If you track this in your file, and especially if the site is open source, you will leak your admin API key. This would mean anyone is able to change anything on your Algolia index.

```shell
// .env.production
ALGOLIA_APP_ID=XXX
ALGOLIA_API_KEY=XXX
ALGOLIA_INDEX_NAME=XXX
```

```js
require('dotenv').config({
  path: `.env.${process.env.NODE_ENV}`,
});

// gatsby-config.js
const myQuery = `
  query {
    pages: allSitePage {
      nodes {
        # querying id is required
        id
        component
        path
        componentChunkName
        jsonName
        internal {
          # querying internal.contentDigest is required
          contentDigest
          type
          owner
        }
      }
    }
  }
`;

const queries = [
  {
    query: myQuery,
    queryVariables: {}, // optional. Allows you to use graphql query variables in the query
    transformer: ({ data }) => data.pages.nodes, // optional
    indexName: 'index name to target', // overrides main index name, optional
    settings: {
      // optional, any index settings
      // Note: by supplying settings, you will overwrite all existing settings on the index
    },
    mergeSettings: false, // optional, defaults to false. See notes on mergeSettings below
  },
];

module.exports = {
  plugins: [
    {
      // This plugin must be placed last in your list of plugins to ensure that it can query all the GraphQL data
      resolve: `gatsby-plugin-algolia`,
      options: {
        appId: process.env.ALGOLIA_APP_ID,
        // Use Admin API key without GATSBY_ prefix, so that the key isn't exposed in the application
        // Tip: use Search API key with GATSBY_ prefix to access the service from within components
        apiKey: process.env.ALGOLIA_API_KEY,
        indexName: process.env.ALGOLIA_INDEX_NAME, // for all queries
        queries,
        chunkSize: 10000, // default: 1000
        settings: {
          // optional, any index settings
          // Note: by supplying settings, you will overwrite all existing settings on the index
        },
        mergeSettings: false, // optional, defaults to false. See notes on mergeSettings below
        concurrentQueries: false, // default: true
        dryRun: false, // default: false, only calculate which objects would be indexed, but do not push to Algolia
        continueOnFailure: false, // default: false, don't fail the build if Algolia indexing fails
        algoliasearchOptions: undefined, // default: { timeouts: { connect: 1, read: 30, write: 30 } }, pass any different options to the algoliasearch constructor
      },
    },
  ],
};
```

The index will be synchronised with the provided index name on Algolia on the `build` step in Gatsby. This is not done earlier to prevent you going over quota while developing.

## Partial Updates

This plugin will update only the changed or deleted nodes on your Gatsby site.

**We rely on Gatsby's default `contentDigest` field, so make sure it is queried.**

## Settings

You can set settings for each index individually (per query), or otherwise it will keep your existing settings.

### Merge Settings

`mergeSettings` allows you to preserve settings changes made on the Algolia website. The default behavior (`mergeSettings: false`) will wipe out your index settings and replace them with settings from the config on each build.

When set to true, the config index settings will be merged with the existing index settings in Algolia (with the config index settings taking precendence).

NOTE: When using `mergeSettings`, any **deleted** settings from the config settings will continue to be persisted since they will still exist in Algolia. If you want to remove a setting, be sure to remove it from both the config and on Algolia's website.

## Concurrent Queries

Sometimes, on limited platforms like Netlify, concurrent queries to the same index can lead to unexpected results or hanging builds. Setting `concurrentQueries` to `false` makes it such that queries are run sequentially rather than concurrently, which may solve some concurrent access issues. Be aware that this option may make indexing take longer than it would otherwise.

## Transformer

The `transformer` field accepts a function and optionally you may provide an `async` function.

## Feedback

This is the very first version of our plugin and isn't yet officially supported. Please leave all your feedback in GitHub issues 😊


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

<a name="1.0.3"></a>
## [1.0.3](https://github.com/algolia/gatsby-plugin-algolia/compare/v1.0.2...v1.0.3) (2023-01-17)


### Bug Fixes

* **multiIndex:** separate cache per index ([#173](https://github.com/algolia/gatsby-plugin-algolia/issues/173)) ([5241366](https://github.com/algolia/gatsby-plugin-algolia/commit/5241366)), closes [#171](https://github.com/algolia/gatsby-plugin-algolia/issues/171)



<a name="1.0.2"></a>
## [1.0.2](https://github.com/algolia/gatsby-plugin-algolia/compare/v1.0.1...v1.0.2) (2022-12-05)


### Bug Fixes

* **importing:** safe check of internal.contentDigest ([bbf1350](https://github.com/algolia/gatsby-plugin-algolia/commit/bbf1350)), closes [#168](https://github.com/algolia/gatsby-plugin-algolia/issues/168)



<a name="1.0.1"></a>
## [1.0.1](https://github.com/algolia/gatsby-plugin-algolia/compare/v1.0.0...v1.0.1) (2022-11-29)


### Bug Fixes

* **initial:** when your index has objects without contentDigest, those should be skipped ([a9c1bd2](https://github.com/algolia/gatsby-plugin-algolia/commit/a9c1bd2)), closes [#167](https://github.com/algolia/gatsby-plugin-algolia/issues/167)



<a name="1.0.0"></a>
# [1.0.0](https://github.com/algolia/gatsby-plugin-algolia/compare/v0.26.0...v1.0.0) (2022-11-28)


### Features

* **dependencies:** support gatsby 5 ([658018e](https://github.com/algolia/gatsby-plugin-algolia/commit/658018e)), closes [#166](https://github.com/algolia/gatsby-plugin-algolia/issues/166)
* **indexing:** simplify and rely on contentDigest ([#162](https://github.com/algolia/gatsby-plugin-algolia/issues/162)) ([4cff643](https://github.com/algolia/gatsby-plugin-algolia/commit/4cff643))


### BREAKING CHANGES

* **indexing:** make enablePartialUpdates true, and remove the option
* **indexing:** rely on contentDigest only for matchFields and remove the option
* **indexing:** remove the skipIndexing flag, use either dryRun or continueOnFailure

For migrating to v1 version smoothly:

1. add `id` instead of `objectID` in your query
2. query for `internal { contentDigest }` and do not transform it
3. set `enablePartialUpdates` to `true`
4. set `matchFields` to `['internal.contentDigest']`

<a name="0.26.0"></a>
# [0.26.0](https://github.com/algolia/gatsby-plugin-algolia/compare/v0.25.0...v0.26.0) (2022-02-03)


### Features

* **queries:** allow variables in graphql query ([4ee890d](https://github.com/algolia/gatsby-plugin-algolia/commit/4ee890d))



<a name="0.25.0"></a>
# [0.25.0](https://github.com/algolia/gatsby-plugin-algolia/compare/v0.24.0...v0.25.0) (2021-12-14)


### Features

* **config:** Allow options to be passed to algoliasearch ([#158](https://github.com/algolia/gatsby-plugin-algolia/issues/158)) ([687ca20](https://github.com/algolia/gatsby-plugin-algolia/commit/687ca20))



<a name="0.24.0"></a>
# [0.24.0](https://github.com/algolia/gatsby-plugin-algolia/compare/v0.23.1...v0.24.0) (2021-11-22)


### Features

* **settings:** allow keeping index settings using mergeSettings ([#153](https://github.com/algolia/gatsby-plugin-algolia/issues/153)) ([362d4a1](https://github.com/algolia/gatsby-plugin-algolia/commit/362d4a1))



<a name="0.23.1"></a>
## [0.23.1](https://github.com/algolia/gatsby-plugin-algolia/compare/v0.23.0...v0.23.1) (2021-11-16)


### Bug Fixes

* use unique temporary indices ([#155](https://github.com/algolia/gatsby-plugin-algolia/issues/155)) ([8916a02](https://github.com/algolia/gatsby-plugin-algolia/commit/8916a02))



<a name="0.23.0"></a>
# [0.23.0](https://github.com/algolia/gatsby-plugin-algolia/compare/v0.22.2...v0.23.0) (2021-11-03)


### Bug Fixes

* removing redundant setSettings call ([#152](https://github.com/algolia/gatsby-plugin-algolia/issues/152)) ([7f872fa](https://github.com/algolia/gatsby-plugin-algolia/commit/7f872fa))


### Features

* **peer:** Allow installation with Gatsby 4 ([#154](https://github.com/algolia/gatsby-plugin-algolia/issues/154)) ([a9477c8](https://github.com/algolia/gatsby-plugin-algolia/commit/a9477c8))



<a name="0.22.2"></a>
## [0.22.2](https://github.com/algolia/gatsby-plugin-algolia/compare/v0.22.1...v0.22.2) (2021-09-10)


### Bug Fixes

* **index:** Creates index if it does not exist ([#151](https://github.com/algolia/gatsby-plugin-algolia/issues/151)) ([8f178ea](https://github.com/algolia/gatsby-plugin-algolia/commit/8f178ea)), closes [#150](https://github.com/algolia/gatsby-plugin-algolia/issues/150)



<a name="0.22.1"></a>
## [0.22.1](https://github.com/algolia/gatsby-plugin-algolia/compare/v0.22.0...v0.22.1) (2021-08-23)


### Bug Fixes

* **matchFields:** don't remove unrelated objects ([cedb355](https://github.com/algolia/gatsby-plugin-algolia/commit/cedb355))



<a name="0.22.0"></a>
# [0.22.0](https://github.com/algolia/gatsby-plugin-algolia/compare/v0.21.0...v0.22.0) (2021-07-12)


### Features

* **continueOnFailure:** continue on env variables missing ([7171a14](https://github.com/algolia/gatsby-plugin-algolia/commit/7171a14)), closes [#141](https://github.com/algolia/gatsby-plugin-algolia/issues/141)



<a name="0.21.0"></a>
# [0.21.0](https://github.com/algolia/gatsby-plugin-algolia/compare/v0.20.1...v0.21.0) (2021-07-08)


### Bug Fixes

* **settings:** forward rules and synonyms to index to write ([ad7e076](https://github.com/algolia/gatsby-plugin-algolia/commit/ad7e076))



<a name="0.20.1"></a>
## [0.20.1](https://github.com/algolia/gatsby-plugin-algolia/compare/v0.20.0...v0.20.1) (2021-05-31)


### Bug Fixes

* **continueOnError:** correct error usage ([8e8c0a1](https://github.com/algolia/gatsby-plugin-algolia/commit/8e8c0a1))



<a name="0.20.0"></a>
# [0.20.0](https://github.com/algolia/gatsby-plugin-algolia/compare/v0.19.0...v0.20.0) (2021-05-18)


### Features


gatsby-node.test.js
const { onPostBuild } = require('./gatsby-node');
const { test, beforeEach, describe } = require('@jest/globals');
const { default: algoliasearch } = require('algoliasearch');

jest.mock('algoliasearch');

const GatsbyGlobals = {
  graphql: jest.fn(),
  reporter: {
    activityTimer: jest.fn(() => ({
      setStatus: jest.fn(),
      start: jest.fn(),
      end: jest.fn(),
    })),
  },
  cache: jest.fn(),
};

beforeEach(() => {
  GatsbyGlobals.graphql.mockClear();
  GatsbyGlobals.reporter.activityTimer.mockClear();
  GatsbyGlobals.cache.mockClear();
});

describe('algoliasearch', () => {
  test('instantiates algoliasearch client', async () => {
    const config = {
      queries: [],
      appId: 'appId',
      apiKey: 'apiKey',
    };
    await onPostBuild(GatsbyGlobals, config);

    expect(algoliasearch).toHaveBeenCalledWith(config.appId, config.apiKey, {
      timeouts: { connect: 1, read: 30, write: 30 },
    });
  });

  test('instantiates algoliasearch client with custom config', async () => {
    const config = {
      queries: [],
      appId: 'appId',
      apiKey: 'apiKey',
      algoliasearchOptions: {
        swag: true,
      },
    };
    await onPostBuild(GatsbyGlobals, config);

    expect(algoliasearch).toHaveBeenCalledWith(config.appId, config.apiKey, {
      swag: true,
    });
  });
});


example/gatsby-config.js
require('dotenv').config({
  path: `.env.${process.env.NODE_ENV}`,
})

const query = `
  {
    allSitePage {
      nodes {
        id
        internal {
          contentDigest
        }
        path
      }
    }
  }
`

const queries = [
  {
    query,
    transformer: ({ data }) => data.allSitePage.nodes,
    // optional
    // indexName: 'pages',
    // optional
    settings: {
      attributesToSnippet: ['path:5'],
    },
  },
]

module.exports = {
  siteMetadata: {
    title: 'Gatsby Algolia Example',
  },
  plugins: [
    {
      // in real life this would be:
      // resolve: 'gatsby-plugin-algolia',
      resolve: require.resolve('../'),
      options: {
        appId: process.env.ALGOLIA_APPID,
        apiKey: process.env.ALGOLIA_APIKEY,
        indexName: process.env.ALGOLIA_INDEXNAME, // for all queries
        queries,
        chunkSize: 10000, // default: 1000
      },
    },
  ],
}


example/src/pages/search.js
import React from 'react'
import algoliasearch from 'algoliasearch/lite'
import {
  Highlight,
  Hits,
  InstantSearch,

  SearchBox,
} from 'react-instantsearch-hooks-web'
import { Link } from 'gatsby'

const searchClient = algoliasearch(
  process.env.GATSBY_ALGOLIA_APPID,
  process.env.GATSBY_ALGOLIA_APIKEY
)

const SearchPage = () => (
  <>
    <Link to="/">Go back to the homepage</Link>

      <InstantSearch
        searchClient={searchClient}
        indexName={process.env.GATSBY_ALGOLIA_INDEXNAME}
      >
        <SearchBox />
        <Hits hitComponent={DefaultHitComponent} />
      </InstantSearch>

  </>
)

const DefaultHitComponent = ({ hit }) => (
  <pre>
    <div>
      id/objectID: <Highlight attribute="id" hit={hit} />,
    </div>
    <div>
      <Link to={hit.path}>
        path: <Highlight attribute="path" hit={hit} />
      </Link>
      ,
    </div>
    <div>
      internal.contentDigest:{' '}
      <Highlight attribute="internal.contentDigest" hit={hit} />
    </div>
  </pre>
)

export default SearchPage


gatsby-node.js
const algoliasearch = require('algoliasearch');
const chunk = require('lodash.chunk');

/**
 * @typedef {import('gatsby').GatsbyCache} GatsbyCache
 */

/**
 * @typedef {import('gatsby').Reporter} Reporter
 */

/**
 * @typedef {import('algoliasearch').SearchIndex} SearchIndex
 */

/**
 * @typedef {import('gatsby').graphql} graphql
 */

/**
 * @typedef {import('@algolia/client-search').Settings} Settings
 */

/**
 * @typedef Query
 * @property {string} query The graphql query
 * @property {object=} queryVariables Allows you to use graphql query variables in the query
 * @property {Function} transformer transform the results of the query into objects. Likely `({ data }) => data.myProperty.nodes`
 * @property {string=} indexName index name for this query
 * @property {Settings} settings index settings for this query
 * @property {boolean=} mergeSettings defaults to false. Whether settings set in the index are overridden or persisted
 */

/**
 * Fetches all records for the current index from Algolia
 * @param {SearchIndex} index
 * @param {Reporter} reporter
 * @param {GatsbyCache} cache
 */
function fetchExistingObjects(index, reporter, cache) {
  const hits = {};

  return cache
    .get(`algolia-objects-${index.indexName}`)
    .then(values => {
      if (!values || Object.keys(values).length === 0) {
        throw new Error('cache actually failed');
      }
      return values;
    })
    .catch(() =>
      index
        .browseObjects({
          batch: batch => {
            if (Array.isArray(batch)) {
              batch.forEach(hit => {
                if (hit.internal?.contentDigest) {
                  hits[hit.objectID] = hit;
                }
              });
            }
          },
          attributesToRetrieve: ['internal.contentDigest'],
        })
        .then(() => hits)
        .catch(err =>
          reporter.panicOnBuild('failed while getting indexed objects', err)
        )
    );
}

/**
 * @typedef PluginConfiguration
 * @property {string} appId
 * @property {string} apiKey
 * @property {import('algoliasearch').AlgoliaSearchOptions} algoliasearchOptions
 * @property {Query[]} queries
 * @property {string} indexName
 * @property {boolean} concurrentQueries
 * @property {boolean} dryRun
 * @property {boolean} continueOnFailure
 */

exports.onPostBuild = async function (
  /** @type {{graphql: graphql, reporter: Reporter, cache: GatsbyCache}} */ {
    graphql,
    reporter,
    cache,
  },
  /** @type {PluginConfiguration} */ config
) {
  const {
    appId,
    apiKey,
    queries,
    concurrentQueries = true,
    dryRun = false,
    continueOnFailure = false,
    algoliasearchOptions = {
      timeouts: {
        connect: 1,
        read: 30,
        write: 30,
      },
    },
  } = config;

  const activity = reporter.activityTimer(`index to Algolia`);
  activity.start();

  if (dryRun === true) {
    console.log(
      '\x1b[33m%s\x1b[0m',
      '==== THIS IS A DRY RUN ====================\n' +
        '- No records will be pushed to your index\n' +
        '- No settings will be updated on your index'
    );
  }

  if (continueOnFailure === true && !(appId && apiKey)) {
    activity.setStatus(
      `options.continueOnFailure is true and api key or appId are missing; skipping indexing`
    );
    activity.end();
    return;
  }

  const client = algoliasearch(appId, apiKey, algoliasearchOptions);

  activity.setStatus(`${queries.length} queries to index`);

  try {
    // combine queries with the same index to prevent overwriting data
    const groupedQueries = groupQueriesByIndex(queries, config);

    const jobs = [];
    for (const [indexName, indexQueries] of Object.entries(groupedQueries)) {
      const queryPromise = runIndexQueries(indexName, indexQueries, {
        client,
        activity,
        graphql,
        cache,
        config,
        reporter,
        dryRun,
      });

      if (concurrentQueries) {
        jobs.push(queryPromise);
      } else {
        // await each individual query rather than batching them
        const res = await queryPromise;
        jobs.push(res);
      }
    }

    await Promise.all(jobs);
  } catch (err) {
    if (continueOnFailure) {
      reporter.warn('failed to index to Algolia');
      console.error(err);
    } else {
      activity.panicOnBuild('failed to index to Algolia', err);
    }
  }

  activity.end();
};

/**
 * @param {PluginConfiguration} config
 */
function groupQueriesByIndex(queries = [], config) {
  const { indexName: mainIndexName } = config;

  return queries.reduce((groupedQueries, queryOptions) => {
    const { indexName = mainIndexName } = queryOptions;

    return {
      ...groupedQueries,
      [indexName]: [
        ...(groupedQueries.hasOwnProperty(indexName)
          ? groupedQueries[indexName]
          : []),
        queryOptions,
      ],
    };
  }, {});
}

/**
 * Run all queries for a given index, then make any updates / removals necessary
 * @param {string} indexName
 * @param {object[]} queries
 * @param {object} options
 * @param {import('algoliasearch').SearchClient} options.client
 * @param {import('gatsby').ActivityTracker} options.activity
 * @param {graphql} options.graphql
 * @param {Reporter} options.reporter
 * @param {PluginConfiguration} options.config
 * @param {GatsbyCache} options.cache
 * @param {boolean=} options.dryRun
 */
async function runIndexQueries(
  indexName,
  queries = [],
  { client, activity, graphql, reporter, cache, config, dryRun }
) {
  const { settings: mainSettings, chunkSize = 1000 } = config;

  activity.setStatus('Getting existing objects');

  const index = await initIndex(client, indexName);

  // get all indexed objects matching all matched fields
  const indexedObjects = await fetchExistingObjects(index, reporter, cache);

  activity.setStatus(
    `Running ${queries.length} ${
      queries.length === 1 ? 'query' : 'queries'
    } for index ${indexName}...`
  );

  const objectMapsByQuery = await Promise.all(
    queries.map(query => getObjectsMapByQuery(query, graphql, reporter))
  );

  const allObjectsMap = objectMapsByQuery.reduce((acc, objectsMap = {}) => {
    return Object.assign(acc, objectsMap);
  }, {});

  cache.set(
    `algolia-objects-${indexName}`,
    Object.fromEntries(
      Object.entries(allObjectsMap).map(([objectID, object]) => [
        objectID,
        { internal: { contentDigest: object.internal?.contentDigest } },
      ])
    )
  );

  activity.setStatus(
    `${queries.length === 1 ? 'Query' : 'Queries'} resulted in a total of ${
      Object.keys(allObjectsMap).length
    } results`
  );

  let toIndex = {}; // used to track objects that should be added / updated
  const toRemove = {}; // used to track objects that are stale and should be removed

  // iterate over each query to determine which data are fresh
  activity.setStatus(`Starting Partial updates...`);

  // iterate over each query
  for (const [i] of queries.entries()) {
    const queryResultsMap = objectMapsByQuery[i] || {};

    // iterate over existing objects and compare to fresh data
    for (const [id, existingObj] of Object.entries(indexedObjects)) {
      if (queryResultsMap.hasOwnProperty(id)) {
        // key matches fresh objects, so compare match fields
        const newObj = queryResultsMap[id];
        if (!newObj.internal?.contentDigest) {
          reporter.panicOnBuild(
            'the objects must have internal.contentDigest. Current object:\n' +
              JSON.stringify(newObj, null, 2)
          );
        }

        if (
          existingObj.internal?.contentDigest !== newObj.internal.contentDigest
        ) {
          // contentDigest differs, so index new object
          toIndex[id] = newObj;
        }
        // objects are the same, so skip

        // remove from queryResultsMap, since it is already accounted for
        delete queryResultsMap[id];
      } else {
        // remove existing object if it is managed and not returned from a query
        if (
          // not in any query
          !allObjectsMap.hasOwnProperty(id) &&
          // managed by this plugin
          existingObj.internal?.contentDigest
        ) {
          toRemove[id] = true;
        }
      }
    }

    if (Object.values(queryResultsMap).length) {
      // stale objects have been removed, remaining query objects should be indexed
      toIndex = {
        ...toIndex,
        ...queryResultsMap,
      };
    }
  }

  const objectsToIndex = Object.values(toIndex);
  const objectsToRemove = Object.keys(toRemove);

  if (objectsToIndex.length) {
    const chunks = chunk(objectsToIndex, chunkSize);

    activity.setStatus(
      `Found ${objectsToIndex.length} new / updated records...`
    );

    if (chunks.length > 1) {
      activity.setStatus(`Splitting in ${chunks.length} jobs`);
    }

    /* Add changed / new objects */
    const chunkJobs = chunks.map(async function (chunked) {
      if (dryRun === true) {
        reporter.info(`[dry run]: ${objectsToIndex.length} records to add`);
      } else {
        await index.saveObjects(chunked);
      }
    });

    await Promise.all(chunkJobs);
  } else {
    activity.setStatus(`No updates necessary; skipping!`);
  }

  if (objectsToRemove.length) {
    activity.setStatus(
      `Found ${objectsToRemove.length} stale objects; removing...`
    );

    if (dryRun === true) {
      reporter.info(`[dry run]: ${objectsToRemove.length} records to delete`);
    } else {
      await index.deleteObjects(objectsToRemove).wait();
    }
  }

  // defer to first query for index settings
  // todo: maybe iterate over all settings and throw if they differ
  const {
    settings = mainSettings,
    mergeSettings = false,
    forwardToReplicas,
  } = queries[0] || {};

  const settingsToApply = await getSettingsToApply({
    settings,
    mergeSettings,
    index,
    reporter,
  });

  if (dryRun) {
    console.log('[dry run]: settings', settingsToApply);
  } else {
    await index
      .setSettings(settingsToApply, {
        forwardToReplicas,
      })
      .wait();
  }

  activity.setStatus('Done!');
}

/**
 * Does an Algolia index exist already
 * @param {import('algoliasearch').SearchIndex} index
 */
function indexExists(index) {
  return index
    .getSettings()
    .then(() => true)
    .catch(error => {
      if (error.status !== 404) {
        throw error;
      }

      return false;
    });
}

/**
 * @param {object} options
 * @param {Settings} options.settings
 * @param {boolean} options.mergeSettings
 * @param {import('algoliasearch').SearchIndex} options.index
 * @param {Reporter} options.reporter
 * @returns {Promise<Settings>}
 */
async function getSettingsToApply({
  settings,
  mergeSettings,
  index,
  reporter,
}) {
  const /** @type Settings */ existingSettings =
      await index.getSettings().catch(e => {
        reporter.panicOnBuild(`${e.toString()} ${index.indexName}`);
      });

  if (!settings) {
    return existingSettings;
  }

  return {
    ...(mergeSettings ? existingSettings : {}),
    ...settings,
  };
}

/**
 * @param {Query} options
 * @param {graphql} graphql
 * @param {Reporter} reporter
 */
async function getObjectsMapByQuery(
  { query, queryVariables, transformer = x => x },
  graphql,
  reporter
) {
  const result = await graphql(query, queryVariables);
  if (result.errors) {
    reporter.panicOnBuild(
      `failed to index to Algolia, errors:\n ${JSON.stringify(result.errors)}`,
      result.errors
    );
  }

  const objects = (await transformer(result)).map(object => ({
    objectID: object.id,
    ...object,
  }));

  if (objects.length > 0 && !objects[0].objectID) {
    reporter.panicOnBuild(
      `failed to index to Algolia. Query does not have 'id'`
    );
  }

  // return a map by id for later use
  return Object.fromEntries(objects.map(object => [object.objectID, object]));
}

/**
 * @param {algoliasearch.SearchClient} client
 * @param {string} indexName
 */
async function initIndex(client, indexName) {
  const index = client.initIndex(indexName);

  if (!(await indexExists(index))) {
    await index.setSettings({}).wait();
  }

  return index;
}
